### PR TITLE
Android platform & binding utilities

### LINKED ISSUE #13: FlowForms android bindings per case
### Requirements :
* Create binding utilities in the FlowForms Android extensions modules to allow the user to bind Views to FlowFields.
* Final usage **in the fragment** should look like this (or similar) : 
```
viewModel.signUpForm.bindFields(
	{ ID1, EmailTextInputView }, 
	{ ID2, RepeatEmailTextInputView }, 
	{ ID3, PasswordTextInputView }
	{ ID4, AcceptTyCCheckBoxView }
)
```
* Update the Android example app to use these utilities instead of manual work.

### Notes : 
* This depends on #8 as there are all the FF's core features.
* Also depends on #12 as there will be created different types of validations to use in this bindings (on value change, on focus change, etc).
* This depends on #5 as it is supposed to be an improvement to the Example app.

### PR DESCRIPTION
**Pull request by hrodrick**
#### ISSUE #13
* closes #13

---

#### Description
* Added binding utilities on FlowForm to bind FlowFields to Views automatically.
* Added `Activity.repeatOnLifeCycleScope(...)` Extension fun to collect flows without all the Lifecycle's boilerPlate
* Added a binding utility to bind FlowFields to liveData changes, so we can use FlowFields with CheckViews and two-way data binding.
* Improve Kdocs
* Add failFast and Async optional params for the Required Validation, true & false by default.
* Changed CI test command to make a check and also trigger android tests.
* Removed FlowForms-Android-Ext as we didn't need it at the end.

---

#### Notes
* Added a TODO to add async validations examples.

### CODE REVIEW

File: ExampleApp Android/src/main/java/com/rootstrap/flowforms/example/MainActivity.kt [hunk 1]
```diff
@@ -35,96 +34,73 @@ class MainActivity : AppCompatActivity() {
...
+            CORRECT, UNMODIFIED -> binding.nameInputLayout.error = null
+            else -> binding.nameInputLayout.error = getString(R.string.required_field)
         }
+    }
 
-        binding.passwordInputEditText.doAfterTextChanged {
-            lifecycleScope.launch {
-                viewModel.form.validateOnValueChange(NEW_PASSWORD)
-            }
+    private fun onEmailStatusChange(status : FieldStatus) {
```
Comment (ximenaperez) [inline ExampleApp Android/src/main/java/com/rootstrap/flowforms/example/MainActivity.kt:74]: same here

File: ExampleApp Android/src/main/java/com/rootstrap/flowforms/example/MainActivity.kt [hunk 2]
```diff
@@ -35,96 +34,73 @@ class MainActivity : AppCompatActivity() {
...
+            INVALID_EMAIL -> binding.emailInputLayout.error = getString(R.string.invalid_email)
+            else -> binding.emailInputLayout.error = getString(R.string.required_field)
         }
+    }
 
-        binding.confirmPasswordInputEditText.doAfterTextChanged {
-            lifecycleScope.launch {
-                viewModel.form.validateOnValueChange(CONFIRM_PASSWORD)
-            }
+    private fun onPasswordStatusChange(status : FieldStatus) {
```
Comment (ximenaperez) [inline ExampleApp Android/src/main/java/com/rootstrap/flowforms/example/MainActivity.kt:82]: same here

File: ExampleApp Android/src/main/java/com/rootstrap/flowforms/example/MainActivity.kt [hunk 3]
```diff
@@ -35,96 +34,73 @@ class MainActivity : AppCompatActivity() {
...
+            MIN_LENGTH_UNSATISFIED -> binding.passwordInputLayout.error = getString(R.string.min_length, MIN_PASSWORD_LENGTH)
+            else -> binding.passwordInputLayout.error = getString(R.string.required_field)
         }
+    }
 
-        viewModel.formModel.confirm.observe(this) {
-            lifecycleScope.launch {
-                viewModel.form.validateOnValueChange(CONFIRMATION)
-            }
+    private fun onConfirmPasswordChange(status : FieldStatus) {
```
Comment (ximenaperez) [inline ExampleApp Android/src/main/java/com/rootstrap/flowforms/example/MainActivity.kt:90]: same here

File: ExampleApp Android/src/main/java/com/rootstrap/flowforms/example/MainActivity.kt [hunk 4]
```diff
@@ -35,96 +34,73 @@ class MainActivity : AppCompatActivity() {
...
+        when (status.code) {
+            CORRECT, UNMODIFIED -> binding.confirmPasswordInputLayout.error = null
+            MIN_LENGTH_UNSATISFIED -> binding.confirmPasswordInputLayout.error = getString(R.string.min_length, MIN_PASSWORD_LENGTH)
+            PASSWORD_MATCH_UNSATISFIED -> binding.confirmPasswordInputLayout.error = getString(R.string.password_match)
+            else -> binding.confirmPasswordInputLayout.error = getString(R.string.required_field)
         }
     }
+
+    private fun onFormStatusChange(status : FormStatus) {
```
Comment (ximenaperez) [inline ExampleApp Android/src/main/java/com/rootstrap/flowforms/example/MainActivity.kt:99]: same here

File: ExampleApp Android/src/main/java/com/rootstrap/flowforms/example/MainActivity.kt [hunk 5]
```diff
@@ -35,96 +34,73 @@ class MainActivity : AppCompatActivity() {
...
+                viewModel.formModel.confirm to CONFIRMATION
+            )
         }
+    }
 
-        binding.emailInputEditText.doAfterTextChanged {
-            lifecycleScope.launch {
-                viewModel.form.validateOnValueChange(EMAIL)
-            }
+    private fun onNameStatusChange(status : FieldStatus) {
```
[the remark this answers is not in the document]
Comment (hrodrick) [reply ExampleApp Android/src/main/java/com/rootstrap/flowforms/example/MainActivity.kt:67]: Yes! Thanks!

File: FlowForms-Core/src/androidMain/kotlin/com/rootstrap/flowforms/util/FlowFormsUtils.kt [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+ *
+ * You can bind any kind of [View], but only those in the following list will be accepted.
+ * * EditText
+ *
+ * @param coroutineScope used to trigger the [FlowField]'s [Validation]s
+ * @param bindPairs a Pair of [View]s and [String]s, each View belonging to each Field in the form
+ * by the paired ID.
+ * @throws [IllegalArgumentException] where there are non-supported views in the pairs.
+ */
+fun FlowForm.bind(coroutineScope : CoroutineScope, vararg bindPairs : Pair<View, String>) {
```
Comment (ximenaperez) [inline FlowForms-Core/src/androidMain/kotlin/com/rootstrap/flowforms/util/FlowFormsUtils.kt:50]: remove the space before :

File: FlowForms-Core/src/androidMain/kotlin/com/rootstrap/flowforms/util/FlowFormsUtils.kt [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+    }
+}
+
+/**
+ * Bind an [EditText] to a [FlowField] ID. Automatically triggering
+ * the corresponding validations (onValueChanged, onBlur, onFocus, etc) when needed.
+ *
+ * @param coroutineScope used to trigger the [FlowField]'s [Validation]s
+ */
+fun FlowForm.bind(coroutineScope : CoroutineScope, editText : EditText, fieldId: String) {
```
Comment (ximenaperez) [inline FlowForms-Core/src/androidMain/kotlin/com/rootstrap/flowforms/util/FlowFormsUtils.kt:65]: same here

File: FlowForms-Core/src/androidMain/kotlin/com/rootstrap/flowforms/util/FlowFormsUtils.kt [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+ *
+ * It only triggers onValueChanged [Validation]s. The [LiveData]'s value is ignored, as the validations
+ * use the field's current value.
+ *
+ * @param lifeCycleOwner used to observe the [LiveData]s
+ * @param coroutineScope used to trigger the [FlowField]'s [Validation]s
+ * @param bindPairs [Pair]s of [LiveData] to [String], each LiveData belonging to each field in the
+ * form by the paired ID
+ */
+fun FlowForm.bind(lifeCycleOwner : LifecycleOwner, coroutineScope : CoroutineScope, vararg bindPairs : Pair<LiveData<*>, String>) {
```
Comment (ximenaperez) [inline FlowForms-Core/src/androidMain/kotlin/com/rootstrap/flowforms/util/FlowFormsUtils.kt:91]: same here

File: FlowForms-Core/src/androidMain/kotlin/com/rootstrap/flowforms/util/OnFocusChangeListenerDecorator.kt [hunk 9]
```diff
@@ -0,0 +1,15 @@
+package com.rootstrap.flowforms.util
+
+import android.view.View
+
+class OnFocusChangeListenerDecorator(
+    private val onFocusChangeListener : View.OnFocusChangeListener?,
```
Comment (ximenaperez) [inline FlowForms-Core/src/androidMain/kotlin/com/rootstrap/flowforms/util/OnFocusChangeListenerDecorator.kt:6]: same here

File: FlowForms-Core/src/androidMain/kotlin/com/rootstrap/flowforms/util/FlowFormsUtils.kt (hunk 6 above)
Comment (hrodrick) [reply FlowForms-Core/src/androidMain/kotlin/com/rootstrap/flowforms/util/FlowFormsUtils.kt:50]: Done, thanks!

File: FlowForms-Core/src/commonMain/kotlin/com/rootstrap/flowforms/core/validation/Required.kt [hunk 10]
```diff
@@ -8,11 +8,17 @@ import com.rootstrap.flowforms.core.common.StatusCodes.REQUIRED_UNSATISFIED
...
+ * Otherwise returns [REQUIRED_UNSATISFIED].
  *
- * @property valueProvider function that returns the value used by the [validate] implementation
+ * @param failFast check failFast property on [Validation] class.
+ * @param async check async property on [Validation] class.
+ * @param valueProvider function that returns the value used by the [validate] implementation.
  */
-class Required(val valueProvider : () -> String?) : Validation() {
+class Required(
+    failFast : Boolean = true,
```
[the remark this answers is not in the document]
Comment (hrodrick) [reply FlowForms-Core/src/commonMain/kotlin/com/rootstrap/flowforms/core/validation/Required.kt:18]: This was fixed in #30 👍🏼

**Review by ximenaperez — APPROVED** [review]

**Review by huanachin — APPROVED** [review]

**Pull request merged**